File: frontend/src/store/projects.ts
import { create } from "zustand";
import { supabase, type Project, type Outline } from "@/lib/supabase";

type ProjectsState = {
  projects: Project[];
  currentProject: Project | null;
  outlines: Outline[];
  isLoading: boolean;
  error: string | null;

  // Actions
  fetchProjects: () => Promise<void>;
  createProject: (title: string, prompt: string, genre?: string) => Promise<Project>;
  selectProject: (id: string) => Promise<void>;
  updateProject: (id: string, updates: Partial<Project>) => Promise<void>;
  fetchOutlines: (projectId: string) => Promise<void>;
  clearError: () => void;
};

export const useProjectsStore = create<ProjectsState>((set, get) => ({
  projects: [],
  currentProject: null,
  outlines: [],
  isLoading: false,
  error: null,

  fetchProjects: async () => {
    set({ isLoading: true, error: null });
    try {
      const { data, error } = await supabase
        .from("morphwrit_projects")
        .select("*")
        .order("updated_at", { ascending: false });

      if (error) throw error;

      set({ projects: data || [], isLoading: false });
    } catch (error) {
      set({ error: (error as Error).message, isLoading: false });
    }
  },

  createProject: async (title: string, prompt: string, genre?: string) => {
    set({ isLoading: true, error: null });
    try {
      const { data: { user } } = await supabase.auth.getUser();
      if (!user) throw new Error("Not authenticated");

      const { data, error } = await supabase
        .from("morphwrit_projects")
        .insert({
          user_id: user.id,
          title,
          prompt,
          genre,
          status: "draft",
        })
        .select()
        .single();

      if (error) throw error;

      set((state) => ({
        projects: [data, ...state.projects],
        currentProject: data,
        isLoading: false,
      }));

      return data;
    } catch (error) {
      set({ error: (error as Error).message, isLoading: false });
      throw error;
    }
  },

  selectProject: async (id: string) => {
    set({ isLoading: true, error: null });
    try {
      const { data, error } = await supabase
        .from("morphwrit_projects")
        .select("*")
        .eq("id", id)
        .single();

      if (error) throw error;

      set({ currentProject: data, isLoading: false });

      // Also fetch outlines
      await get().fetchOutlines(id);
    } catch (error) {
      set({ error: (error as Error).message, isLoading: false });
    }
  },

  updateProject: async (id: string, updates: Partial<Project>) => {
    try {
      const { data, error } = await supabase
        .from("morphwrit_projects")
        .update(updates)
        .eq("id", id)
        .select()
        .single();

      if (error) throw error;

      set((state) => ({
        projects: state.projects.map((p) => (p.id === id ? data : p)),
        currentProject: state.currentProject?.id === id ? data : state.currentProject,
      }));
    } catch (error) {
      set({ error: (error as Error).message });
    }
  },

  fetchOutlines: async (projectId: string) => {
    try {
      const { data, error } = await supabase
        .from("morphwrit_outlines")
        .select("*")
        .eq("project_id", projectId)
        .order("layer");

      if (error) throw error;

      set({ outlines: data || [] });
    } catch (error) {
      set({ error: (error as Error).message });
    }
  },

  clearError: () => set({ error: null }),
}));

File: backend/api/projects.py
"""
Projects API Routes
"""

from fastapi import APIRouter, HTTPException, Depends
from typing import List

from backend.models.project import (
    ProjectCreate, ProjectUpdate, ProjectResponse,
    OutlineCreate, OutlineResponse,
    DraftCreate, DraftResponse,
)
from backend.api.deps import get_current_user_id
from backend.core.supabase import get_supabase_client
from backend.core.logging import get_logger

router = APIRouter()
logger = get_logger("api.projects")


@router.get("/", response_model=List[ProjectResponse])
async def list_projects(user_id: str = Depends(get_current_user_id)):
    """List all projects for current user."""
    try:
        client = get_supabase_client()
        response = client.table("morphwrit_projects") \
            .select("*") \
            .eq("user_id", user_id) \
            .order("updated_at", desc=True) \
            .execute()

        return response.data or []

    except Exception as e:
        logger.error(f"List projects error: {e}")
        raise HTTPException(status_code=500, detail="Failed to list projects")


@router.post("/", response_model=ProjectResponse)
async def create_project(
    project: ProjectCreate,
    user_id: str = Depends(get_current_user_id)
):
    """Create a new project."""
    try:
        client = get_supabase_client()
        response = client.table("morphwrit_projects").insert({
            "user_id": user_id,
            "title": project.title,
            "prompt": project.prompt,
            "genre": project.genre,
            "status": "draft",
            "world_config": project.world_config or {}
        }).execute()

        if not response.data:
            raise HTTPException(status_code=400, detail="Failed to create project")

        return response.data[0]

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Create project error: {e}")
        raise HTTPException(status_code=500, detail="Failed to create project")


@router.get("/{project_id}", response_model=ProjectResponse)
async def get_project(
    project_id: str,
    user_id: str = Depends(get_current_user_id)
):
    """Get a specific project."""
    try:
        client = get_supabase_client()
        response = client.table("morphwrit_projects") \
            .select("*") \
            .eq("id", project_id) \
            .eq("user_id", user_id) \
            .single() \
            .execute()

        if not response.data:
            raise HTTPException(status_code=404, detail="Project not found")

        return response.data

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Get project error: {e}")
        raise HTTPException(status_code=500, detail="Failed to get project")


@router.patch("/{project_id}", response_model=ProjectResponse)
async def update_project(
    project_id: str,
    updates: ProjectUpdate,
    user_id: str = Depends(get_current_user_id)
):
    """Update a project."""
    try:
        client = get_supabase_client()

        update_data = updates.model_dump(exclude_unset=True)
        if not update_data:
            raise HTTPException(status_code=400, detail="No updates provided")

        response = client.table("morphwrit_projects") \
            .update(update_data) \
            .eq("id", project_id) \
            .eq("user_id", user_id) \
            .execute()

        if not response.data:
            raise HTTPException(status_code=404, detail="Project not found")

        return response.data[0]

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Update project error: {e}")
        raise HTTPException(status_code=500, detail="Failed to update project")


# Outlines endpoints
@router.get("/{project_id}/outlines", response_model=List[OutlineResponse])
async def get_outlines(
    project_id: str,
    user_id: str = Depends(get_current_user_id)
):
    """Get all outlines for a project."""
    try:
        # Verify project ownership first
        client = get_supabase_client()
        project = client.table("morphwrit_projects") \
            .select("id") \
            .eq("id", project_id) \
            .eq("user_id", user_id) \
            .single() \
            .execute()

        if not project.data:
            raise HTTPException(status_code=404, detail="Project not found")

        response = client.table("morphwrit_outlines") \
            .select("*") \
            .eq("project_id", project_id) \
            .order("layer") \
            .execute()

        return response.data or []

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Get outlines error: {e}")
        raise HTTPException(status_code=500, detail="Failed to get outlines")

